File: bridge/ts/src/contract.ts
/**
 * Edge-plane projection of bridge/contract/turn.schema.json.
 *
 * The JSON Schema is canonical. `contract.parity.test.ts` fails the build if
 * this file drifts from it, so "single source of truth" is enforced rather
 * than merely asserted in a comment.
 */

export const CONTRACT_VERSION = "turn.v1" as const;

export interface TraceContext {
  /** 16-byte hex, W3C traceparent trace-id. One trace spans BOTH planes. */
  trace_id: string;
  /** 8-byte hex; the edge-side span that originated the turn. */
  span_id: string;
  sampled?: boolean;
}

export type TargetKind = "agent" | "team" | "workflow";

export interface Target {
  kind: TargetKind;
  id: string;
}

export interface Attachment {
  kind: "image" | "audio" | "video" | "file";
  uri: string;
  mime_type?: string;
  size_bytes?: number;
}

export interface Principal {
  user_id: string;
  display_name?: string;
  channel: string;
  channel_user?: string;
  /** The brain plane ENFORCES these, not merely logs them. */
  scopes?: string[];
}

export interface TurnRequest {
  /** Idempotency key. Redelivery must not execute twice. */
  turn_id: string;
  session_id: string;
  target: Target;
  input: { text: string; attachments?: Attachment[] };
  principal: Principal;
  trace: TraceContext;
  options?: { stream?: boolean; timeout_ms?: number };
}

export interface Usage {
  input_tokens?: number;
  output_tokens?: number;
  tool_calls?: number;
  duration_ms?: number;
}

export type TurnEvent =
  | { type: "run.started"; turn_id: string; run_id?: string; ts: string }
  | { type: "output.delta"; turn_id: string; text: string; ts: string }
  | { type: "reasoning.delta"; turn_id: string; text: string; ts: string }
  | { type: "tool.started"; turn_id: string; tool: string; args_preview?: string; ts: string }
  | { type: "tool.completed"; turn_id: string; tool: string; ok: boolean; error?: string; ts: string }
  /** PAUSE (REQ-0042): the run waits for a decision until `expires_at`, then is BLOCKED. */
  | { type: "approval.required"; turn_id: string; approval_id: string; prompt: string; tool?: string; expires_at?: string; ts: string }
  | { type: "run.completed"; turn_id: string; run_id?: string; output: string; usage?: Usage; ts: string }
  /** Terminal. A BLOCKED run is `run.failed` with `retryable: false` and a `reason` in BLOCK_REASONS. */
  | { type: "run.failed"; turn_id: string; error: string; retryable?: boolean; reason?: FailureReason; ts: string };

/** Why a run ended without completing; mirrors the schema's `reason` enum on run.failed. */
export type FailureReason =
  | "approval_timed_out"
  | "approval_denied"
  | "kernel_denied"
  | "kernel_unreachable"
  | "pause_not_resumable"
  | "executor_error"
  | "client_disconnected"
  | "cancelled";

/** Runtime mirror of FailureReason, checked against the schema by contract.parity.test.ts. */
export const FAILURE_REASONS = [
  "approval_timed_out",
  "approval_denied",
  "kernel_denied",
  "kernel_unreachable",
  "pause_not_resumable",
  "executor_error",
  "client_disconnected",
  "cancelled",
] as const satisfies readonly FailureReason[];

/** The reasons that mean BLOCK (REQ-0042): a tool was refused or an approval could not be obtained. */
export const BLOCK_REASONS = [
  "approval_timed_out",
  "approval_denied",
  "kernel_denied",
  "kernel_unreachable",
  "pause_not_resumable",
] as const satisfies readonly FailureReason[];

export function isBlocked(ev: TurnEvent): boolean {
  return ev.type === "run.failed" && ev.reason !== undefined && (BLOCK_REASONS as readonly string[]).includes(ev.reason);
}

export type TurnEventType = TurnEvent["type"];

export const TERMINAL_EVENT_TYPES = ["run.completed", "run.failed"] as const;

export function isTerminal(ev: TurnEvent): ev is Extract<TurnEvent, { type: "run.completed" | "run.failed" }> {
  return ev.type === "run.completed" || ev.type === "run.failed";
}

export interface ApprovalDecision {
  approval_id: string;
  decision: "approve" | "deny";
  principal: Principal;
  reason?: string;
}

/** Generates a W3C-shaped trace context for a turn originating at the edge. */
export function newTraceContext(): TraceContext {
  const hex = (bytes: number) =>
    Array.from(crypto.getRandomValues(new Uint8Array(bytes)))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  return { trace_id: hex(16), span_id: hex(8), sampled: true };
}

/**
 * Runtime-visible list of every TurnEvent discriminator.
 *
 * TypeScript types are erased at runtime, so the union above cannot be compared
 * against the canonical JSON Schema by a test. This array is the runtime mirror,
 * and the two assertions below make it impossible to drift from the union
 * WITHOUT tsc failing:
 *
 *   - `_AllUnionMembersListed` fails if a union member is missing from the array.
 *   - `_NoExtraMembersListed`  fails if the array names something not in the union.
 *
 * `contract.parity.test.ts` then checks this array against turn.schema.json, so
 * schema -> array -> union is closed end to end.
 */
export const TURN_EVENT_TYPES = [
  "run.started",
  "output.delta",
  "reasoning.delta",
  "tool.started",
  "tool.completed",
  "approval.required",
  "run.completed",
  "run.failed",
] as const;

type _AllUnionMembersListed =
  Exclude<TurnEventType, (typeof TURN_EVENT_TYPES)[number]> extends never ? true : never;
type _NoExtraMembersListed =
  Exclude<(typeof TURN_EVENT_TYPES)[number], TurnEventType> extends never ? true : never;

const _unionCoverage: _AllUnionMembersListed = true;
const _arrayCoverage: _NoExtraMembersListed = true;
void _unionCoverage;
void _arrayCoverage;

File: bridge/py/agenticos_bridge/contract.py
"""Typed mirror of bridge/contract/turn.schema.json.

The JSON Schema is canonical. This module is the Python projection of it, and
`test_contract_parity` asserts the two never drift. Illegal states are made
unrepresentable here (discriminated unions, constrained strings, forbidden
extras) so a malformed turn is rejected at the boundary rather than halfway
through an agent run.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Annotated, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field

TRACE_ID_RE = r"^[0-9a-f]{32}$"
SPAN_ID_RE = r"^[0-9a-f]{16}$"


def _utcnow() -> str:
    return datetime.now(timezone.utc).isoformat()


class _Strict(BaseModel):
    """Reject unknown fields at the plane boundary rather than silently dropping them."""

    model_config = ConfigDict(extra="forbid", frozen=True)


class TraceContext(_Strict):
    trace_id: str = Field(pattern=TRACE_ID_RE)
    span_id: str = Field(pattern=SPAN_ID_RE)
    sampled: bool = True


class Target(_Strict):
    kind: Literal["agent", "team", "workflow"]
    id: str = Field(min_length=1, max_length=128)


class Attachment(_Strict):
    kind: Literal["image", "audio", "video", "file"]
    uri: str
    mime_type: Optional[str] = None
    size_bytes: Optional[int] = Field(default=None, ge=0)


class Principal(_Strict):
    user_id: str = Field(min_length=1)
    display_name: Optional[str] = None
    channel: str
    channel_user: Optional[str] = None
    scopes: list[str] = Field(default_factory=list)


class TurnInput(_Strict):
    text: str
    attachments: list[Attachment] = Field(default_factory=list, max_length=32)


class TurnOptions(_Strict):
    stream: bool = True
    timeout_ms: int = Field(default=120_000, ge=1_000, le=3_600_000)


class TurnRequest(_Strict):
    turn_id: str = Field(min_length=1)
    session_id: str = Field(min_length=1)
    target: Target
    input: TurnInput
    principal: Principal
    trace: TraceContext
    options: TurnOptions = Field(default_factory=TurnOptions)


class Usage(_Strict):
    input_tokens: int = Field(default=0, ge=0)
    output_tokens: int = Field(default=0, ge=0)
    tool_calls: int = Field(default=0, ge=0)
    duration_ms: int = Field(default=0, ge=0)


class RunStarted(_Strict):
    type: Literal["run.started"] = "run.started"
    turn_id: str
    run_id: Optional[str] = None
    ts: str = Field(default_factory=_utcnow)


class OutputDelta(_Strict):
    type: Literal["output.delta"] = "output.delta"
    turn_id: str
    text: str
    ts: str = Field(default_factory=_utcnow)


class ReasoningDelta(_Strict):
    """Brain-plane reasoning trace.

    Agno emits ReasoningStarted/ReasoningStep/ReasoningContentDelta/ReasoningCompleted.
    Without this member the bridge would discard the brain plane's strongest
    differentiator at the boundary. The edge plane may decline to *render* these
    in a channel, but must still receive them so traces stay complete.
    """

    type: Literal["reasoning.delta"] = "reasoning.delta"
    turn_id: str
    text: str
    ts: str = Field(default_factory=_utcnow)


class ToolStarted(_Strict):
    type: Literal["tool.started"] = "tool.started"
    turn_id: str
    tool: str
    args_preview: Optional[str] = Field(default=None, max_length=512)
    ts: str = Field(default_factory=_utcnow)


class ToolCompleted(_Strict):
    type: Literal["tool.completed"] = "tool.completed"
    turn_id: str
    tool: str
    ok: bool
    error: Optional[str] = None
    ts: str = Field(default_factory=_utcnow)


class ApprovalRequired(_Strict):
    """PAUSE (REQ-0042): the run is waiting for a decision.

    ``expires_at`` is the instant the pause lapses. Until then the brain
    waits and resumes on approve; past it the run is BLOCKED (``RunFailed``
    with ``reason="approval_timed_out"``) and, under kernel governance, the
    kernel's own approval record is resolved as revoked rather than merely
    left to lapse by TTL.
    """

    type: Literal["approval.required"] = "approval.required"
    turn_id: str
    approval_id: str
    prompt: str
    tool: Optional[str] = None
    expires_at: Optional[str] = None
    ts: str = Field(default_factory=_utcnow)


class RunCompleted(_Strict):
    type: Literal["run.completed"] = "run.completed"
    turn_id: str
    run_id: Optional[str] = None
    output: str
    usage: Usage = Field(default_factory=Usage)
    ts: str = Field(default_factory=_utcnow)


FailureReason = Literal[
    "approval_timed_out",
    "approval_denied",
    "kernel_denied",
    "kernel_unreachable",
    "pause_not_resumable",
    "executor_error",
    "client_disconnected",
    "cancelled",
]

# The reasons that mean BLOCK (REQ-0042): a tool call was refused, or an
# approval could not be obtained, and the run was ended rather than resumed.
BLOCK_REASONS = frozenset(
    {"approval_timed_out", "approval_denied", "kernel_denied", "kernel_unreachable", "pause_not_resumable"}
)


class RunFailed(_Strict):
    """Terminal. A BLOCKED run is a ``RunFailed`` with ``retryable=False`` and
    a ``reason`` in ``BLOCK_REASONS``; ``error`` stays the human-readable text."""

    type: Literal["run.failed"] = "run.failed"
    turn_id: str
    error: str
    retryable: bool = False
    reason: Optional[FailureReason] = None
    ts: str = Field(default_factory=_utcnow)


TurnEvent = Annotated[
    Union[
        RunStarted,
        OutputDelta,
        ReasoningDelta,
        ToolStarted,
        ToolCompleted,
        ApprovalRequired,
        RunCompleted,
        RunFailed,
    ],
    Field(discriminator="type"),
]

TERMINAL_EVENT_TYPES = frozenset({"run.completed", "run.failed"})


class ApprovalDecision(_Strict):
    approval_id: str
    decision: Literal["approve", "deny"]
    principal: Principal
    reason: Optional[str] = Field(default=None, max_length=1024)


__all__ = [
    "TraceContext", "Target", "Attachment", "Principal", "TurnInput",
    "TurnOptions", "TurnRequest", "Usage", "RunStarted", "OutputDelta",
    "ReasoningDelta", "ToolStarted", "ToolCompleted", "ApprovalRequired", "RunCompleted",
    "RunFailed", "FailureReason", "BLOCK_REASONS", "TurnEvent", "ApprovalDecision",
    "TERMINAL_EVENT_TYPES",
]
